### PR TITLE
Feature window x11

### PR DESCRIPTION
**Pull request by Quentin01**

### CODE REVIEW

File: samples/window/main.cpp [hunk 1]
```diff
@@ -15,11 +15,7 @@ int main() {
             if (event.type == lug::Window::EventType::CLOSE) {
                 window->close();
             }
-
```
Comment (Quentin01) [inline samples/window/main.cpp]: Why have you modified this file?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 2]
```diff
@@ -1,11 +1,120 @@
+#include <queue>
+#include <cstring>
+#include <string>
 #include <lug/Window/Unix/WindowImplX11.hpp>
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: This include had to be first

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 3]
```diff
@@ -1,11 +1,120 @@
...
+#include <lug/Window/Unix/WmHints.hpp>
 
-bool lug::Window::priv::WindowImpl::create(uint16_t width, uint16_t height, const std::string& title, Style style) {
-    return false;
+
+lug::Window::priv::WindowImpl::WindowImpl(Window* win): _parent{win} {}
+
+bool lug::Window::priv::WindowImpl::create(const std::string& title, Style style) {
+    uint32_t blackColor = 0;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: We are not in C, declare your variables when you use them.

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 4]
```diff
@@ -1,11 +1,120 @@
...
+    uint8_t screen = 0;
+    GC _graphic_context;
+
+    _display = XOpenDisplay(nullptr);
+    if (_display == nullptr) {
+            return false;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Indentation

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 5]
```diff
@@ -1,11 +1,120 @@
...
+    _graphic_context = DefaultGC(_display, screen);
+    ::Window parent = RootWindow(_display, screen);
+
+    blackColor = BlackPixel(_display, DefaultScreen(_display));
+
+    _window = XCreateSimpleWindow(_display, parent, 0, 0, _parent->_mode.width, _parent->_mode.height, 90, 2, blackColor);
+
+    if (!_window) {
+        XCloseDisplay(_display);
+        return false;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Set `_display` to `nullptr`

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 6]
```diff
@@ -1,11 +1,120 @@
...
 
-bool lug::Window::priv::WindowImpl::create(uint16_t width, uint16_t height, const std::string& title, Style style) {
-    return false;
+
+lug::Window::priv::WindowImpl::WindowImpl(Window* win): _parent{win} {}
+
+bool lug::Window::priv::WindowImpl::create(const std::string& title, Style style) {
+    uint32_t blackColor = 0;
+    uint8_t screen = 0;
+    GC _graphic_context;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Why this variable have an underscore?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 7]
```diff
@@ -1,11 +1,120 @@
...
+    if (_display != nullptr) {
+        XCloseDisplay(_display);
+        XDestroyWindow(_display, _window);
+        _display = nullptr;
+        _window = 0;
+    }
 }
 
-void lug::Window::priv::WindowImpl::processEvents(std::queue<lug::Window::Event>& events) {
+Bool selectEvents(Display* display, XEvent* event, XPointer arg)
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: `bool`?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 8]
```diff
@@ -1,11 +1,120 @@
...
+{
+    if (event->type == ClientMessage || event->type == DestroyNotify)
+        return True;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: `true`?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 9]
```diff
@@ -1,11 +1,120 @@
...
+    else
+        return False;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: `false`?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 10]
```diff
@@ -1,11 +1,120 @@
...
+        XDestroyWindow(_display, _window);
+        _display = nullptr;
+        _window = 0;
+    }
 }
 
-void lug::Window::priv::WindowImpl::processEvents(std::queue<lug::Window::Event>& events) {
+Bool selectEvents(Display* display, XEvent* event, XPointer arg)
+{
+    if (event->type == ClientMessage || event->type == DestroyNotify)
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Brackets?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 11]
```diff
@@ -1,11 +1,120 @@
...
+        return True;
+    else
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Brackets?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 12]
```diff
@@ -1,11 +1,120 @@
...
+        XCloseDisplay(_display);
+        XDestroyWindow(_display, _window);
+        _display = nullptr;
+        _window = 0;
+    }
 }
 
-void lug::Window::priv::WindowImpl::processEvents(std::queue<lug::Window::Event>& events) {
+Bool selectEvents(Display* display, XEvent* event, XPointer arg)
+{
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Bracket at the end of the previous line

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 13]
```diff
@@ -1,11 +1,120 @@
...
+        return True;
+    else
+        return False;
 }
+
+bool lug::Window::priv::WindowImpl::pollEvent(lug::Window::Event& event) {
+    XEvent xEvent;
+
+    if (XCheckIfEvent(_display, &xEvent, selectEvents, nullptr) == True)
+    {
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Bracket at the end of the previous line

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 14]
```diff
@@ -1,11 +1,120 @@
...
+    if (event->type == ClientMessage || event->type == DestroyNotify)
+        return True;
+    else
+        return False;
 }
+
+bool lug::Window::priv::WindowImpl::pollEvent(lug::Window::Event& event) {
+    XEvent xEvent;
+
+    if (XCheckIfEvent(_display, &xEvent, selectEvents, nullptr) == True)
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: `true`?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 15]
```diff
@@ -1,11 +1,120 @@
...
+                }
+                break;
+            case DestroyNotify:
+                event.type = lug::Window::EventType::DESTROY;
+                break;
+            default:
+                return false;
+        }
+    }
+    else
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Brackets

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 16]
```diff
@@ -1,11 +1,120 @@
...
+        return false;
+
+    return true;
+}
+
+void lug::Window::priv::WindowImpl::setWindowDecorations(Style style)
+{
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Bracket at the end of the previous line

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 17]
```diff
@@ -1,11 +1,120 @@
...
+    WMHints hints;
+    std::memset(&hints, 0, sizeof(hints));
+    hints.flags = MWM_HINTS_FUNCTIONS | MWM_HINTS_DECORATIONS;
+    hints.decorations = 0;
+    hints.functions = 0;
+
+    if (static_cast<bool>(style & Style::Titlebar))
+    {
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Bracket at the end of the previous line

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 18]
```diff
@@ -1,11 +1,120 @@
...
+        hints.decorations |= MWM_DECOR_BORDER | MWM_DECOR_TITLE | MWM_DECOR_MINIMIZE | MWM_DECOR_MENU;
+        hints.functions   |= MWM_FUNC_MOVE | MWM_FUNC_MINIMIZE;
+    }
+    if (static_cast<bool>(style & Style::Resize))
+    {
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Bracket at the end of the previous line

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 19]
```diff
@@ -1,11 +1,120 @@
...
+        hints.decorations |= MWM_DECOR_MAXIMIZE | MWM_DECOR_RESIZEH;
+        hints.functions   |= MWM_FUNC_MAXIMIZE | MWM_FUNC_RESIZE;
+    }
+    if (static_cast<bool>(style & Style::Close))
+    {
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Bracket at the end of the previous line

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 20]
```diff
@@ -1,11 +1,120 @@
...
+
+    XChangeProperty(_display,
+                    _window,
+                    _wmHints,
+                    _wmHints,
+                    32,
+                    PropModeReplace,
+                    reinterpret_cast<const unsigned char*>(&hints),
+                    5);
+}
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: New line at the end of the file

File: src/lug/Window/Unix/WindowImplX11.hpp [hunk 21]
```diff
@@ -1,24 +1,39 @@
...
+
+    bool pollEvent(lug::Window::Event& event);
+
+private:
+    void setWindowDecorations(Style style);
+
+private:
+    Window* _parent{nullptr};
+
+	Display* _display;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.hpp]: Indentation and initialize it

File: src/lug/Window/Unix/WindowImplX11.hpp [hunk 22]
```diff
@@ -1,24 +1,39 @@
...
+	::Window  _window;
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.hpp]: Indentation and initialize it

File: src/lug/Window/Unix/WmHints.hpp [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+
+#define MWM_FUNC_RESIZE         (1L << 1)
+#define MWM_FUNC_MOVE           (1L << 2)
+#define MWM_FUNC_MINIMIZE       (1L << 3)
+#define MWM_FUNC_MAXIMIZE       (1L << 4)
+
+#define MWM_FUNC_CLOSE          (1L << 5)
+
+struct WMHints
+{
```
Comment (Quentin01) [inline src/lug/Window/Unix/WmHints.hpp]: Bracket at the end of the previous line

File: src/lug/Window/Window.cpp [hunk 24]
```diff
@@ -1,3 +1,6 @@
+#include <memory>
```
Comment (Quentin01) [inline src/lug/Window/Window.cpp]: Why are you addind these includes?

**Review by Quentin01 — COMMENTED** [review]
Request changes

File: src/lug/Window/Unix/WmHints.hpp [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+    unsigned long flags;
+    unsigned long functions;
+    unsigned long decorations;
+    long          inputMode;
+    unsigned long state;
+};
+
+} // namespace priv
+} // namespace Window
+} // namespace lug
```
Comment (saveman71) [inline src/lug/Window/Unix/WmHints.hpp]: Add a newline at end of file

File: src/lug/Window/Unix/WindowImplX11.cpp (hunk 14 above)
Comment (saveman71) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Invert the condition, so you can return false earlier and lose an indentation level

File: src/lug/Window/Unix/WmHints.hpp [hunk 26]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+namespace lug {
+namespace Window {
+namespace priv {
+
+#define MWM_HINTS_FUNCTIONS     (1L << 0)
```
Comment (saveman71) [inline src/lug/Window/Unix/WmHints.hpp]: Use `constexpr` instead

`constexpr uint8_t MWMHintsFunctions = 1 << 0`

Same for all the defines below

**Review by saveman71 — CHANGES_REQUESTED** [review]

File: samples/window/main.cpp (hunk 1 above)
Comment (mapy69003) [inline samples/window/main.cpp]: For tests

Comment (saveman71) [reply samples/window/main.cpp]: What he meant is do not push these useless modifications

Comment (yoannpicquenot) [reply samples/window/main.cpp]: Yup man

File: src/lug/Window/Unix/WmHints.hpp (hunk 25 above)
Comment (mapy69003) [inline src/lug/Window/Unix/WmHints.hpp]: I already add a newline, but the comment still staying ?

Comment (saveman71) [inline src/lug/Window/Unix/WmHints.hpp]: Probably a bug from Github

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 27]
```diff
@@ -64,47 +62,41 @@ Bool selectEvents(Display* display, XEvent* event, XPointer arg)
...
-                break;
-            case DestroyNotify:
-                event.type = lug::Window::EventType::DESTROY;
-                break;
-            default:
-                return false;
-        }
-    }
-    else
+    if (XCheckIfEvent(_display, &xEvent, selectEvents, nullptr) == False) {
```
Comment (saveman71) [inline src/lug/Window/Unix/WindowImplX11.cpp]: `false`

Comment (saveman71) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Or simply use `!` before `XCheckIfEvent`

File: src/lug/Window/Window.cpp [hunk 28]
```diff
@@ -1,4 +1,3 @@
-#include <memory>
 #include <string>
```
Comment (saveman71) [inline src/lug/Window/Window.cpp]: These are not useful too?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 29]
```diff
@@ -64,47 +62,41 @@ Bool selectEvents(Display* display, XEvent* event, XPointer arg)
...
         return false;
+    }
+    switch (xEvent.type) {
```
Comment (saveman71) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Add a newline before the switch for more clarity

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 30]
```diff
@@ -1,42 +1,41 @@
...
 
     _display = XOpenDisplay(nullptr);
     if (_display == nullptr) {
-            return false;
+        return false;
     }
 
-    screen = DefaultScreen(_display);
-    _graphic_context = DefaultGC(_display, screen);
+    uint8_t screen = DefaultScreen(_display);
```
Comment (saveman71) [inline src/lug/Window/Unix/WindowImplX11.cpp]: DefaultScreen returns an `int`

**Review by saveman71 — CHANGES_REQUESTED** [review]

**Comment by yoannpicquenot** [comment]
saveman71: Yes, You can

**Review by saveman71 — APPROVED** [review]

**Pull request merged**